### PR TITLE
provider: Support arbitrary additional tag data, implement keyvaluetags and ignore_tags support in aws_autoscaling_group resource, return empty strings with key-only tags instead of panicking

### LINKED ISSUE #12368: r/aws_lightsail_instance: Crash when adding tag with empty value
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

AWS provider v2.52.0.

### Affected Resource(s)

* `aws_lightsail_instance`

### Terraform Configuration Files

```hcl
resource "aws_lightsail_instance" "lightsail_instance_test" {
  name              = "terraform-test"
  availability_zone = "${data.aws_availability_zones.available.names[0]}"
  blueprint_id      = "amazon_linux"
  bundle_id         = "nano_1_0"

  tags = {
    Name = "terraform-test"
    KeyOnlyTag = ""
  }
}
```

### Panic Output

```
aws_lightsail_instance.lightsail_instance_test: Still creating... (10s elapsed)
aws_lightsail_instance.lightsail_instance_test: Still creating... (20s elapsed)

Error: Error applying plan:

1 error occurred:
	* aws_lightsail_instance.lightsail_instance_test: 1 error occurred:
	* aws_lightsail_instance.lightsail_instance_test: unexpected EOF





Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.


panic: runtime error: invalid memory address or nil pointer dereference
2020-03-12T09:27:00.257-0400 [DEBUG] plugin.terraform-provider-aws: [signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x384f82a]
2020-03-12T09:27:00.257-0400 [DEBUG] plugin.terraform-provider-aws: 
2020-03-12T09:27:00.257-0400 [DEBUG] plugin.terraform-provider-aws: goroutine 257 [running]:
2020-03-12T09:27:00.257-0400 [DEBUG] plugin.terraform-provider-aws: github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags.KeyValueTags.Map(0xc000a679b0, 0xc000a679b0)
2020-03-12T09:27:00.257-0400 [DEBUG] plugin.terraform-provider-aws: 	/home/kit/wrk/src/github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags/key_value_tags.go:138 +0xda
2020-03-12T09:27:00.257-0400 [DEBUG] plugin.terraform-provider-aws: github.com/terraform-providers/terraform-provider-aws/aws.resourceAwsLightsailInstanceRead(0xc00039e150, 0x4df1960, 0xc000ee2500, 0x0, 0x0)
2020-03-12T09:27:00.257-0400 [DEBUG] plugin.terraform-provider-aws: 	/home/kit/wrk/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_lightsail_instance.go:216 +0x6eb
2020-03-12T09:27:00.257-0400 [DEBUG] plugin.terraform-provider-aws: github.com/terraform-providers/terraform-provider-aws/aws.resourceAwsLightsailInstanceCreate(0xc00039e150, 0x4df1960, 0xc000ee2500, 0xc00039e150, 0x0)
2020-03-12T09:27:00.258-0400 [DEBUG] plugin.terraform-provider-aws: 	/home/kit/wrk/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_lightsail_instance.go:170 +0x723
2020-03-12T09:27:00.258-0400 [DEBUG] plugin.terraform-provider-aws: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).Apply(0xc000a34b00, 0xc0001cd860, 0xc0005c1a60, 0x4df1960, 0xc000ee2500, 0xc0006ce001, 0xc000075ba8, 0x40da86)
2020-03-12T09:27:00.258-0400 [DEBUG] plugin.terraform-provider-aws: 	/home/kit/wrk/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.7.0/helper/schema/resource.go:305 +0x365
2020-03-12T09:27:00.258-0400 [DEBUG] plugin.terraform-provider-aws: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Provider).Apply(0xc000c19480, 0xc000bb8040, 0xc0001cd860, 0xc0005c1a60, 0xc000d15a70, 0xc000ce7ef0, 0x18)
2020-03-12T09:27:00.258-0400 [DEBUG] plugin.terraform-provider-aws: 	/home/kit/wrk/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.7.0/helper/schema/provider.go:294 +0x99
2020-03-12T09:27:00.258-0400 [DEBUG] plugin.terraform-provider-aws: github.com/hashicorp/terraform-plugin-sdk/plugin.(*ResourceProviderServer).Apply(0xc00052bd20, 0xc0005c14a0, 0xc000530400, 0x0, 0x0)
2020-03-12T09:27:00.258-0400 [DEBUG] plugin.terraform-provider-aws: 	/home/kit/wrk/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.7.0/plugin/resource_provider.go:530 +0x57
2020-03-12T09:27:00.258-0400 [DEBUG] plugin.terraform-provider-aws: reflect.Value.call(0xc000550cc0, 0xc000bcdfe8, 0x13, 0x58b4395, 0x4, 0xc000075f18, 0x3, 0x3, 0x406e9a, 0xc0003b7560, ...)
2020-03-12T09:27:00.258-0400 [DEBUG] plugin.terraform-provider-aws: 	/usr/local/go/src/reflect/value.go:460 +0x5f6
2020-03-12T09:27:00.258-0400 [DEBUG] plugin.terraform-provider-aws: reflect.Value.Call(0xc000550cc0, 0xc000bcdfe8, 0x13, 0xc000846718, 0x3, 0x3, 0x67569b, 0xc000ff8000, 0x0)
2020-03-12T09:27:00.258-0400 [DEBUG] plugin.terraform-provider-aws: 	/usr/local/go/src/reflect/value.go:321 +0xb4
2020-03-12T09:27:00.258-0400 [DEBUG] plugin.terraform-provider-aws: net/rpc.(*service).call(0xc000c9a8c0, 0xc0005179f0, 0xc0004ca0b0, 0xc0004ca0c0, 0xc000c19500, 0xc0000c1460, 0x460e900, 0xc0005c14a0, 0x16, 0x460e940, ...)
2020-03-12T09:27:00.258-0400 [DEBUG] plugin.terraform-provider-aws: 	/usr/local/go/src/net/rpc/server.go:377 +0x16f
2020-03-12T09:27:00.258-0400 [DEBUG] plugin.terraform-provider-aws: created by net/rpc.(*Server).ServeCodec
2020-03-12T09:27:00.258-0400 [DEBUG] plugin.terraform-provider-aws: 	/usr/local/go/src/net/rpc/server.go:474 +0x42b
```

### Steps to Reproduce

1. `terraform apply`

### References

First noticed by @stefansundin in https://github.com/terraform-providers/terraform-provider-aws/pull/12346.

Relevant code:

https://github.com/terraform-providers/terraform-provider-aws/blob/654cc23a38444d4836155d6a2243e3976b5f3992/aws/internal/keyvaluetags/key_value_tags.go#L134-L142

### LINKED ISSUE #12701: Terraform crashing when creating bucket in MinIO
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform v0.11.7

### Terraform Configuration Files

```hcl
provider "aws" {
  region                      = "us-east-1"
  skip_credentials_validation = true
  skip_requesting_account_id  = true
  s3_force_path_style         = true

  endpoints {
    s3 = "https://minio.example.com"
  }
}

resource "aws_s3_bucket" "b" {
  bucket = "my-tf-test-bucket"
}

```

### Panic Output

https://gist.github.com/aboutte/0532c669df1d66c411f35eb9eadc33fa

### Expected Behavior

- terraform creates the bucket
- bucket gets stored in state

### Actual Behavior

- terraform creates the bucket
- terraform crashes when receiving response from Minio

### Steps to Reproduce

- deploy minio
- review the steps and terraform provided in the gist above

### LINKED ISSUE #13808: Support aws_autoscaling_group in keyvaluetags and with Provider ignore_tags Configuration
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

All AWS services support resource tags with just a key and value, except Auto Scaling, where there is an additional `PropagateAtLaunch` field. The `keyvaluetags` package was deployed with provider `ignore_tags` support for all Terraform AWS Provider resources except `aws_autoscaling_group` due to this quirk. This issue serves to backfill that support into the last resource.

The first step will be adding "additional data" support into the `keyvaluetags` package functions, adding the `autoscaling` service name to all `keyvaluetags` generators, switching the `aws_autoscaling_group` resource `tag` attribute implementation to the generated handling, and adding ignore configuration support to the resource.

### Affected Resource(s)

* aws_autoscaling_group

### Potential Terraform Configuration

```hcl
provider "aws" {
  # ... potentially other configuration ...
  ignore_tags {
    keys = ["AmazonECSManaged"]
    key_prefixes = ["kubernetes.io/"]
  }
}
```

### References

- https://www.terraform.io/docs/providers/aws/#ignore_tags
- https://github.com/terraform-providers/terraform-provider-aws/issues/10688
- https://github.com/terraform-providers/terraform-provider-aws/issues/10689
- https://github.com/terraform-providers/terraform-provider-aws/pull/13039

### LINKED ISSUE #14768: Autoscaling group tags are destroyed and recreated for an update in-place
_This issue was originally opened by @Lynty as hashicorp/terraform#25949. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._
    
<hr>
    

### Terraform Version

Hello,
I have a Service Control Policy configured in an AWS account that denies creation and updating of an autoscaling group if it does not have some mandatory tags. I am using this AWS [example](https://docs.aws.amazon.com/organizations/latest/userguide/orgs_manage_policies_example-scps.html#example-require-tag-on-create) to accomplish this. It uses a Null condition operator to check if a condition key is present at the time of authorization. If it is `true` (the key doesn't exist — it is null), the SCP prevents resource creation.

The issue I am coming across is that regardless of a valid or invalid change to a tag, a `terraform plan` output shows a change in place for the autoscaling group resource but actually destroys and recreates the changed tag. A `terraform apply` will trigger the SCP and deny updating the autoscaling group due to the tag being momentarily destroyed/marked as `null`.

```
Terraform v0.12.24
+ provider.aws v2.52.0
```

### Terraform Configuration Files

```terraform
resource "aws_autoscaling_group" "webserver" {
  name_prefix               = "webserver"
  max_size                  = 1
  min_size                  = 0
  desired_capacity          = 1
  health_check_grace_period = 300
  health_check_type         = "EC2"
  force_delete              = true

  lifecycle {
    create_before_destroy = true
  }

  launch_configuration = aws_launch_configuration.webserver.name
  # this will become the private_subnets once we have a bastion
  vpc_zone_identifier = var.public_subnets

  tag {
    key                 = "environment"
    value               = var.env
    propagate_at_launch = "true"
  }
  tag {
    key                 = "project"
    value               = "mvp"
    propagate_at_launch = "true"
  }
  tag {
    key                 = "Name"
    value               = "${terraform.workspace}-${var.env}-webserver"
    propagate_at_launch = "true"
  }
  tag {
    key                 = "workspace"
    value               = terraform.workspace
    propagate_at_launch = "true"
  }
  tag {
    key                 = "component"
    value               = "webserver"
    propagate_at_launch = "true"
  }
  tag {
    key                 = "terraform"
    value               = "True"
    propagate_at_launch = "true"
  }
}

```
Service Control Policy Document
```
data "aws_iam_policy_document" "scp_mandatory_ec2_tags" {
  version = "2012-10-17"
  statement {
    sid = "DenyAutoscalingGroupWithNoEnvironmentTag"

    actions = [
      "autoscaling:CreateAutoScalingGroup",
    ]

    effect = "Deny"

    resources = [
      "arn:aws:autoscaling:*:*:autoScalingGroup:*:autoScalingGroupName/*",
    ]

    condition {
      test     = "Null"
      variable = "aws:RequestTag/environment"

      values = [
        "true",
      ]
    }
  }

  statement {
    sid = "DenyAutoscalingGroupWithNoProjectTag"

    actions = [
      "autoscaling:CreateAutoScalingGroup",
    ]

    effect = "Deny"

    resources = [
      "arn:aws:autoscaling:*:*:autoScalingGroup:*:autoScalingGroupName/*",
    ]

    condition {
      test     = "Null"
      variable = "aws:RequestTag/project"

      values = [
        "true",
      ]
    }
  }

  statement {
    sid = "DenyAutoscalingGroupWithNoNameTag"

    actions = [
      "autoscaling:CreateAutoScalingGroup",
    ]

    effect = "Deny"

    resources = [
      "arn:aws:autoscaling:*:*:autoScalingGroup:*:autoScalingGroupName/*",
    ]

    condition {
      test     = "Null"
      variable = "aws:RequestTag/Name"

      values = [
        "true",
      ]
    }
  }

  statement {
    sid = "DenyAutoscalingGroupWithNoComponentTag"

    actions = [
      "autoscaling:CreateAutoScalingGroup",
    ]

    effect = "Deny"

    resources = [
      "arn:aws:autoscaling:*:*:autoScalingGroup:*:autoScalingGroupName/*",
    ]

    condition {
      test     = "Null"
      variable = "aws:RequestTag/component"

      values = [
        "true",
      ]
    }
  }

  statement {
    sid = "DenyAutoscalingGroupWithNoWorkspaceTag"

    actions = [
      "autoscaling:CreateAutoScalingGroup",
    ]

    effect = "Deny"

    resources = [
      "arn:aws:autoscaling:*:*:autoScalingGroup:*:autoScalingGroupName/*",
    ]

    condition {
      test     = "Null"
      variable = "aws:RequestTag/workspace"

      values = [
        "true",
      ]
    }
  }

  statement {
    sid = "DenyAutoscalingGroupWithNoGxPTag"

    actions = [
      "autoscaling:CreateAutoScalingGroup",
    ]

    effect = "Deny"

    resources = [
      "arn:aws:autoscaling:*:*:autoScalingGroup:*:autoScalingGroupName/*",
    ]

    condition {
      test     = "Null"
      variable = "aws:RequestTag/gxp"

      values = [
        "true",
      ]
    }
  }

  statement {
    sid = "DenyAutoscalingGroupWithNoTerraformTag"

    actions = [
      "autoscaling:CreateAutoScalingGroup",
    ]

    effect = "Deny"

    resources = [
      "arn:aws:autoscaling:*:*:autoScalingGroup:*:autoScalingGroupName/*",
    ]

    condition {
      test     = "Null"
      variable = "aws:RequestTag/terraform"

      values = [
        "true",
      ]
    }
  }

  statement {
    sid = "DenyAutoScalingGroupWithoutGxPBoolean"

    actions = [
      "autoscaling:CreateAutoScalingGroup",
    ]

    effect = "Deny"

    resources = [
      "arn:aws:autoscaling:*:*:autoScalingGroup:*:autoScalingGroupName/*",
    ]

    condition {
      test     = "StringNotEquals"
      variable = "aws:RequestTag/gxp"

      values = [
        "True",
        "False"
      ]
    }
  }

  statement {
    sid = "DenyAutoScalingGroupWithoutTerraformBoolean"

    actions = [
      "autoscaling:CreateAutoScalingGroup",
    ]

    effect = "Deny"

    resources = [
      "arn:aws:autoscaling:*:*:autoScalingGroup:*:autoScalingGroupName/*",
    ]

    condition {
      test     = "StringNotEquals"
      variable = "aws:RequestTag/terraform"

      values = [
        "True",
        "False"
      ]
    }
  }
}

```

### Expected Behavior

`environment` is a mandatory tag. I should spin up an ec2 instance in an autoscaling group with tag `environment:dev`. While the instance is running, I should be able to change the tag value to something like `environment:stage` without triggering the Service Control Policy with terraform.

### Actual Behavior

SCP is triggered and blocks autoscaling resource update due to seeing a tag being destroyed/marked as `null`. 
`terraform plan`
```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

aws_autoscaling_group.webserver: Refreshing state... [id=webserver20200820173423399100000001]

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # aws_autoscaling_group.webserver will be updated in-place
  ~ resource "aws_autoscaling_group" "webserver" {
        arn                       = "arn:aws:autoscaling:us-west-2:104568625881:autoScalingGroup:624943a4-2ab9-4585-90ca-c30a185c14d3:autoScalingGroupName/webserver20200820173423399100000001"
        availability_zones        = [
            "us-west-2a",
            "us-west-2b",
        ]
        default_cooldown          = 300
        desired_capacity          = 1
        enabled_metrics           = []
        force_delete              = true
        health_check_grace_period = 300
        health_check_type         = "EC2"
        id                        = "webserver20200820173423399100000001"
        launch_configuration      = "webserver20200819192444336200000002"
        load_balancers            = []
        max_instance_lifetime     = 0
        max_size                  = 1
        metrics_granularity       = "1Minute"
        min_size                  = 0
        name                      = "webserver20200820173423399100000001"
        name_prefix               = "webserver"
        protect_from_scale_in     = false
        service_linked_role_arn   = "arn:aws:iam::104568625881:role/aws-service-role/autoscaling.amazonaws.com/AWSServiceRoleForAutoScaling"
        suspended_processes       = []
        target_group_arns         = []
        termination_policies      = []
        vpc_zone_identifier       = [
            "subnet-0cac79a2fb0d14a70",
            "subnet-0d9d17f1e0c6f5eba",
        ]
        wait_for_capacity_timeout = "10m"

        tag {
            key                 = "Name"
            propagate_at_launch = true
            value               = "ldong5-opsdev-webserver"
        }
        tag {
            key                 = "component"
            propagate_at_launch = true
            value               = "webserver"
        }
      - tag {
          - key                 = "environment" -> null
          - propagate_at_launch = true -> null
          - value               = "opsdev" -> null
        }
      + tag {
          + key                 = "environment"
          + propagate_at_launch = true
          + value               = "test"
        }
        tag {
            key                 = "gxp"
            propagate_at_launch = true
            value               = "False"
        }
        tag {
            key                 = "project"
            propagate_at_launch = true
            value               = "mvp"
        }
        tag {
            key                 = "terraform"
            propagate_at_launch = true
            value               = "True"
        }
        tag {
            key                 = "workspace"
            propagate_at_launch = true
            value               = "ldong5"
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.

------------------------------------------------------------------------

Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.
```
### Steps to Reproduce

1. `terraform init`
2. `terraform apply`
3. Change a tag value to a different valid value
4. `terraform apply`

### PR DESCRIPTION
**Pull request by bflad**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #12368
Closes #12701
Closes #13808
Closes #14768

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
* resource/aws_autoscaling_group: Support provider-level `ignore_tags` configuration
* resource/aws_lightsail_instance: Prevent panic with key-only tags
```

Output from acceptance testing:

```
--- PASS: TestAccAWSAutoScalingGroup_ALB_TargetGroups (270.49s)
--- PASS: TestAccAWSAutoScalingGroup_ALB_TargetGroups_ELBCapacity (347.98s)
--- PASS: TestAccAWSAutoScalingGroup_autoGeneratedName (48.39s)
--- PASS: TestAccAWSAutoScalingGroup_basic (240.73s)
--- PASS: TestAccAWSAutoScalingGroup_classicVpcZoneIdentifier (57.40s)
--- PASS: TestAccAWSAutoScalingGroup_emptyAvailabilityZones (301.11s)
--- PASS: TestAccAWSAutoScalingGroup_enablingMetrics (203.91s)
--- PASS: TestAccAWSAutoScalingGroup_initialLifecycleHook (262.41s)
--- PASS: TestAccAWSAutoScalingGroup_launchTemplate (80.07s)
--- PASS: TestAccAWSAutoScalingGroup_LaunchTemplate_IAMInstanceProfile (100.28s)
--- PASS: TestAccAWSAutoScalingGroup_launchTemplate_update (141.65s)
--- PASS: TestAccAWSAutoScalingGroup_launchTempPartitionNum (80.33s)
--- PASS: TestAccAWSAutoScalingGroup_LoadBalancers (835.18s)
--- PASS: TestAccAWSAutoScalingGroup_MaxInstanceLifetime (108.58s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy (54.75s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_OnDemandAllocationStrategy (83.27s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_OnDemandBaseCapacity (86.71s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_OnDemandPercentageAboveBaseCapacity (84.96s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_SpotAllocationStrategy (47.10s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_SpotInstancePools (89.58s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_SpotMaxPrice (85.66s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_UpdateToZeroOnDemandBaseCapacity (75.63s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_LaunchTemplate_LaunchTemplateSpecification_LaunchTemplateName (48.09s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_LaunchTemplate_LaunchTemplateSpecification_Version (87.27s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_LaunchTemplate_Override_InstanceType (90.33s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_LaunchTemplate_Override_WeightedCapacity (177.12s)
--- PASS: TestAccAWSAutoScalingGroup_namePrefix (77.41s)
--- PASS: TestAccAWSAutoScalingGroup_serviceLinkedRoleARN (43.51s)
--- PASS: TestAccAWSAutoScalingGroup_suspendingProcesses (210.03s)
--- PASS: TestAccAWSAutoScalingGroup_tags (228.44s)
--- PASS: TestAccAWSAutoScalingGroup_TargetGroupArns (243.95s)
--- PASS: TestAccAWSAutoScalingGroup_terminationPolicies (173.57s)
--- PASS: TestAccAWSAutoScalingGroup_VpcUpdates (81.11s)
--- PASS: TestAccAWSAutoScalingGroup_WithLoadBalancer (315.22s)
--- PASS: TestAccAWSAutoScalingGroup_WithLoadBalancer_ToTargetGroup (316.01s)
--- PASS: TestAccAWSAutoScalingGroup_withMetrics (115.03s)
--- PASS: TestAccAWSAutoScalingGroup_withPlacementGroup (232.21s)
```

### CODE REVIEW

**Comment by ewbankkit** [comment]
Supercedes:
* https://github.com/terraform-providers/terraform-provider-aws/pull/3533

**Comment by bflad** [comment]
Rebased and re-verified:

```
--- PASS: TestAccAWSAutoScalingGroup_ALB_TargetGroups (175.59s)
--- PASS: TestAccAWSAutoScalingGroup_ALB_TargetGroups_ELBCapacity (307.41s)
--- PASS: TestAccAWSAutoScalingGroup_autoGeneratedName (53.58s)
--- PASS: TestAccAWSAutoScalingGroup_basic (271.50s)
--- PASS: TestAccAWSAutoScalingGroup_classicVpcZoneIdentifier (81.34s)
--- PASS: TestAccAWSAutoScalingGroup_emptyAvailabilityZones (63.89s)
--- PASS: TestAccAWSAutoScalingGroup_enablingMetrics (158.10s)
--- PASS: TestAccAWSAutoScalingGroup_initialLifecycleHook (272.15s)
--- PASS: TestAccAWSAutoScalingGroup_launchTemplate (51.46s)
--- PASS: TestAccAWSAutoScalingGroup_LaunchTemplate_IAMInstanceProfile (59.35s)
--- PASS: TestAccAWSAutoScalingGroup_launchTemplate_update (172.15s)
--- PASS: TestAccAWSAutoScalingGroup_launchTempPartitionNum (53.72s)
--- PASS: TestAccAWSAutoScalingGroup_LoadBalancers (556.01s)
--- PASS: TestAccAWSAutoScalingGroup_MaxInstanceLifetime (97.18s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy (84.80s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_OnDemandAllocationStrategy (84.68s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_OnDemandBaseCapacity (142.47s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_OnDemandPercentageAboveBaseCapacity (110.84s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_SpotAllocationStrategy (57.34s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_SpotInstancePools (84.39s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_SpotMaxPrice (147.12s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_InstancesDistribution_UpdateToZeroOnDemandBaseCapacity (89.85s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_LaunchTemplate_LaunchTemplateSpecification_LaunchTemplateName (48.93s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_LaunchTemplate_LaunchTemplateSpecification_Version (83.02s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_LaunchTemplate_Override_InstanceType (146.64s)
--- PASS: TestAccAWSAutoScalingGroup_MixedInstancesPolicy_LaunchTemplate_Override_WeightedCapacity (169.50s)
--- PASS: TestAccAWSAutoScalingGroup_namePrefix (91.83s)
--- PASS: TestAccAWSAutoScalingGroup_serviceLinkedRoleARN (81.65s)
--- PASS: TestAccAWSAutoScalingGroup_suspendingProcesses (235.15s)
--- PASS: TestAccAWSAutoScalingGroup_tags (244.95s)
--- PASS: TestAccAWSAutoScalingGroup_TargetGroupArns (239.74s)
--- PASS: TestAccAWSAutoScalingGroup_terminationPolicies (141.07s)
--- PASS: TestAccAWSAutoScalingGroup_VpcUpdates (174.64s)
--- PASS: TestAccAWSAutoScalingGroup_WithLoadBalancer (388.90s)
--- PASS: TestAccAWSAutoScalingGroup_WithLoadBalancer_ToTargetGroup (361.88s)
--- PASS: TestAccAWSAutoScalingGroup_withMetrics (78.39s)
--- PASS: TestAccAWSAutoScalingGroup_withPlacementGroup (182.13s)
```

File: aws/internal/keyvaluetags/generators/createtags/main.go [hunk 1]
```diff
@@ -133,25 +134,27 @@ func isResourceTimeoutError(err error) bool {
 // {{ . | Title }}CreateTags creates {{ . }} service tags for new resources.
```
Comment (ewbankkit) [inline aws/internal/keyvaluetags/generators/createtags/main.go:134]: `isAWSErrCode()` and `isAWSErrCodeContains()` above can also be removed now and replaced by functions in `github.com/hashicorp/aws-sdk-go-base/tfawserr`.

**Comment by ewbankkit** [comment]
Is there an opportunity to address https://github.com/terraform-providers/terraform-provider-aws/issues/11480 with this one, modifying `internal/keyvaluetags/generators/updatetags/main.go`, changing `updatedTags := oldTags.Updated(newTags)` to `updatedTags := oldTags.Updated(newTags).IgnoreAws()` (and the same for `removedTags`)?

**Comment by bflad** [comment]
The changes necessary for #11480 are not related to the intentions of this pull request, but a followup can certainly be submitted with covering acceptance test(s). 👍

**Pull request merged**